arangod/RocksDBEngine/RocksDBRateLimiterThread.cpp

```diff
@@ -0,0 +1,375 @@
+          if (compactionSlots > 0) {
+            pendingCompactionBytes = totalCompactionDebt / compactionSlots;
+          }
+
+          auto internalRocksDB = dynamic_cast<rocksdb::DBImpl*>(_db);
+          TRI_ASSERT(internalRocksDB != nullptr);
+
+          double percentReached = 0.0;
+          if (uint64_t compactionHardLimit =
+                  internalRocksDB->GetOptions()
+                      .hard_pending_compaction_bytes_limit;
+              compactionHardLimit > 0) {
+            // if we are above 25% of the pending compaction bytes stop trigger,
+            // take everything into account that is above this threshold, and
+            // use it to slow down the writes.
+            uint64_t threshold = compactionHardLimit / 4;
+            if (pendingCompactionBytes > threshold) {
+              // we are above the threshold, so penalize writes so that
+              // compaction can keep up long-term. the closer we are to the stop
+              // trigger, the more we will subtract from the target write rate.
+              percentReached = std::min<double>(
+                  0.99,
+                  static_cast<double>(pendingCompactionBytes - threshold) /
+                      static_cast<double>(compactionHardLimit - threshold));
+              targetRate -= percentReached * targetRate;
+            }
+          }
+
+          // fetch our old write rate
+          uint64_t oldRate = _currentRate.load(std::memory_order_relaxed);
+          // calculate the new write rate. to reduce volatility, the new
+          // write rate will be based on the old write rate, and the
+          // difference between target write rate and old write rate is
+          // only applied gradually (using a scaling factor). the larger
+          // the scaling factor is, the smoother the write rate adjustments
+          // will be, but the slower the reaction to changes will be.
+          uint64_t newRate = std::invoke([&]() {
+            if (oldRate == 0) {
+              // never had a write rate set. use our initially calculated
+              // target write rate as the starting point.
+              return targetRate;
+            }
+            if (targetRate > oldRate) {
+              // increase write rate. for this used a reduced scaling factor
+              // (scaling factor / 2), so that increases in the write rate
+              // will be propagated more quickly than reductions.
+              return oldRate + (targetRate - oldRate) / (_scalingFactor / 2);
+            } else {
+              // decreate write rate. for this use the original scaling factor.
+              return oldRate - (oldRate - targetRate) / _scalingFactor;
+            }
+          });
+
+          // write rate must always be between configured minimum and maximum
+          // write rates.
+          newRate = std::clamp(newRate, _minWriteRate, _maxWriteRate);
+
+          LOG_TOPIC("37e36", INFO, Logger::ENGINES)
```

Comment:
Note to self: log level needs to be lowered to `DEBUG` here. The `INFO` is currently only there to have enough output during some performance test runs.